File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return DFS(root, LONG_MIN, LONG_MAX);
    }
    bool DFS(TreeNode* root, long left, long right)
    {
        if(!root) return true;
        if (!(left < root->val && root->val < right)) return false;
        return DFS(root->left, left, root->val) and 
               DFS(root->right, root->val, right);

    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {

    public bool IsValidBST(TreeNode root) {
        return DFS(root, long.MinValue, long.MaxValue);
    }

    public bool DFS(TreeNode root, long left, long right)
    {
        if (root == null) return true;

        if (!(left < root.val && root.val < right)) return false;

        return DFS(root.left, left, root.val) && DFS(root.right, root.val, right);
    }
}